### PR TITLE
Quantized Linear

### PR DESCRIPTION
**Pull request by mmoffatt2**
Moved quantize linear code from "Quantization and Binarization Implementations for Linear Layers" to new PR

### CODE REVIEW

File: gpt_conf.py [hunk 1]
```diff
@@ -114,9 +114,19 @@ class GPTConfig:
...
+    linear_variant_mlp_up: str = None
+    linear_variant_mlp_down: str = None
 
     ## Linear Initialization Options
     linear_mean_init: float= 0.0
     linear_std_init: float= 0.02
 
+    # Quantization Options
+    quantization_linear_method: str = "affine_quant"
+    quantization_linear_bits: int = 8
```
Comment (gkielian) [inline gpt_conf.py:132]: It will be a lot more lines, but let's add ability to quantize the outputs of each of the linear matrices separately.

File: model.py [hunk 2]
```diff
@@ -97,9 +101,21 @@ def __init__(self, config, fire_pos_enc=None):
...
+        # if v variant is specified, takes precedence over linear_variant_attn
+        if config.linear_variant_v:
+            self.c_attn_v = linear_dictionary[config.linear_variant_v](config.n_embd, self.kv_dim, config=config, bias=config.bias)
+        else:
+            self.c_attn_v = linear_dictionary[config.linear_variant_attn](config.n_embd, self.kv_dim, config=config, bias=config.bias)
+        if config.linear_variant_attn_proj:
+            self.c_proj = linear_dictionary[config.linear_variant_attn_proj](config.n_embd, config.n_embd, config=config, bias=config.bias)
+        else:
+            self.c_proj = linear_dictionary[config.linear_variant_attn](config.n_embd, config.n_embd, config=config, bias=config.bias)
+
```
Comment (gkielian) [inline model.py:118]: Let's think how we can reduce the if thens here, maybe some strategy with a dictionary, or an additional boolean flag for uniform attn precision

File: model.py [hunk 3]
```diff
@@ -265,12 +278,25 @@ def __init__(self, config):
...
-            self.c_fc = linear_dictionary[config.linear_variant](config.n_embd, 4 * config.n_embd, config=config, bias=config.bias)
-            self.c_proj = linear_dictionary[config.linear_variant](4 * config.n_embd, config.n_embd, config=config, bias=config.bias)
+            if config.linear_variant_mlp_up:
+                self.c_fc = linear_dictionary[config.linear_variant_mlp_up](config.n_embd, 4 * config.n_embd, config=config, bias=config.bias)
+            else:
+                self.c_fc = linear_dictionary[config.linear_variant_mlp](config.n_embd, 4 * config.n_embd, config=config, bias=config.bias)
+            if config.linear_variant_mlp_down:
+                self.c_proj = linear_dictionary[config.linear_variant_mlp_down](4 * config.n_embd, config.n_embd, config=config, bias=config.bias)
+            else:
+                self.c_proj = linear_dictionary[config.linear_variant_mlp](4 * config.n_embd, config.n_embd, config=config, bias=config.bias)
```
Comment (gkielian) [inline model.py:288]: Same, hoping we can reduce the if elses

File: model.py [hunk 4]
```diff
@@ -265,12 +278,25 @@ def __init__(self, config):
...
+            if config.linear_variant_mlp_up:
+                self.c_fc_in1 = linear_dictionary[config.linear_variant_mlp_up](config.n_embd, 4 * config.n_embd, config=config)
+                self.c_fc_in2 = linear_dictionary[config.linear_variant_mlp_up](config.n_embd, 4 * config.n_embd, config=config)
+            else:
+                self.c_fc_in1 = linear_dictionary[config.linear_variant_mlp](config.n_embd, 4 * config.n_embd, config=config)
+                self.c_fc_in2 = linear_dictionary[config.linear_variant_mlp](config.n_embd, 4 * config.n_embd, config=config)
+            if config.linear_variant_mlp_down:
+                self.c_fc_out = linear_dictionary[config.linear_variant_mlp_down](4 * config.n_embd, config.n_embd, config=config)
+            else:
+                self.c_fc_out = linear_dictionary[config.linear_variant_mlp](4 * config.n_embd, config.n_embd, config=config)
```
Comment (gkielian) [inline model.py:299]: Same

**Comment by djlisbonne** [comment]
Does this also add QAT / is there a cmd line flag for that?

File: train.py [hunk 5]
```diff
@@ -111,7 +111,7 @@ def parse_args():
 
     # LINEAR VARIATIONS
     model_group.add_argument(
-        "--linear_variant",
+        "--linear_variant_attn",
```
Comment (gkielian) [inline train.py:114]: I think this intersects another PR, I'll get back to you on this one.

File: variations/linear_variations.py [hunk 6]
```diff
@@ -4,12 +4,96 @@
...
+    def __init__(self, in_features, out_features, config, bias=True, warmup_step=0):
+        super().__init__(in_features, out_features, bias)
+
+        self.config = config
+        self.weight_bits = config.quantization_linear_bits
+
+        if self.weight_bits < 1:
+            raise ValueError(f"weight_bits={self.weight_bits} must be higher than 0 ")
+
+        self.warmup_step = warmup_step
```
Comment (gkielian) [inline variations/linear_variations.py:29]: There should be a config for this -- in fact is this tied to warmup iterations?

File: variations/linear_variations.py [hunk 7]
```diff
@@ -4,12 +4,96 @@
...
     def __init__(self, in_features, out_features, config=None, bias=None):
         super(WrappedLinear, self).__init__(in_features, out_features, bias)
 
+class QuantizedLinear(nn.Linear):
+    """Linear layer with quantization aware training capability
+    Source: https://github.com/Alexstrasza98/Transformer-Quantization/blob/main
+    Source License: MIT
+    """
+
+    def __init__(self, in_features, out_features, config, bias=True, warmup_step=0):
```
Comment (gkielian) [inline variations/linear_variations.py:20]: Should we tie the warmup_step to warmup_iters?

**Review by gkielian — CHANGES_REQUESTED** [review]
Let's think about ways to prevent the if else statements, must be something we can do.

**Comment by mmoffatt2** [comment]
> Does this also add QAT / is there a cmd line flag for that?

Currently, this PR only supports QAT

File: train.py (hunk 5 above)
Comment (mmoffatt2) [reply train.py:114]: I saw this was in the "Add granular rmsnorm linear settings so that we can and bitlinear with RMSNorm" PR, so I copied the same format.

File: variations/linear_variations.py (hunk 6 above)
Comment (mmoffatt2) [reply variations/linear_variations.py]: Not sure if we want to directly tie this to warmup_iters. For now, I added a quantization_warmup_iters config argument.

File: train.py [hunk 8]
```diff
@@ -126,19 +126,132 @@ def parse_args():
...
+        "--linear_variant_q",
+        type=str,
+        default=None,
+        choices=[
+            "linear",
+            "bitlinear",
+            "bitlinear_1p58",
+            "bitlinear_optimized",
+            "kan",
+            "quantized_linear"
```
Comment (gkielian) [inline train.py:142]: Since these are shared, we can create a linear_variant list and use for each of the linear variant argparses

File: train.py [hunk 9]
```diff
@@ -126,19 +126,132 @@ def parse_args():
...
     model_group.add_argument("--quantize_wte", default=None, action=argparse.BooleanOptionalAction, help="Whether the word embedding is quantized")
     model_group.add_argument("--quantize_wpe", default=None, action=argparse.BooleanOptionalAction, help="Whether the word position embedding is quantized")
-    model_group.add_argument("--quantization_wte_method", type=str, default="affine_quant", choices=["affine_quant", "stochastic_quant"], help="function used for word embedding quantization")
-    model_group.add_argument("--quantization_wte_bits", type=int, default=8, help="number of bits for word embedding quantization")
-    model_group.add_argument("--quantization_wpe_method", type=str, default="affine_quant", choices=["affine_quant", "stochastic_quant"], help="function used for position embedding quantization")
-    model_group.add_argument("--quantization_wpe_bits", type=int, default=8, help="number of bits for position embedding quantization")
+    model_group.add_argument("--quantize_wte_method", type=str, default="affine_quant", choices=["symmetric_quant", "affine_quant", "stochastic_quant"], help="function used for word embedding quantization")
```
Comment (gkielian) [inline train.py:236]: Same here, we can create a choice list for each of these, to help prevent massive updates.

**Review by gkielian — CHANGES_REQUESTED** [review]
Have a few ideas how to further min the code.

Going to start working through the quantization section tomorrow

**Review by gkielian — APPROVED** [review]
